keps/sig-node/5474-enable-writable-cgroups/README.md

```diff
@@ -0,0 +1,615 @@
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes adding a `WritableCgroups` field to the container SecurityContext in Kubernetes to allow unprivileged containers to have writable access to cgroup interfaces on cgroup v2 systems.
+
+## Motivation
+
+ With cgroup v2's secure delegation model, unprivileged containers can safely manage their own cgroup subtree without compromising system security. To support this, a configuration option can be introduced to ensure that, when cgroup v2 is enabled, the cgroup interface (/sys/fs/cgroup) is mounted with read-write permissions for containers.
+
+By exposing the `WritableCgroups` field through the Kubernetes API and CRI interface, container runtimes can be updated to honor the setting via CRI, enabling unprivileged containers to take advantage of writable cgroups in a secure manner.
+
+Related Issues:
+- https://github.com/containerd/containerd/issues/10924
+- https://github.com/kubernetes/kubernetes/issues/121190
+
+### Goals
+
+- Add a `WritableCgroups` boolean field to the container SecurityContext object
+- CRI API changes
+- Integrate with Pod Security Standards to ensure appropriate security policy enforcement
+- Container runtime modifications
+- Maintain backward compatibility with existing workloads
+
+## Proposal
+
+Add a new `WritableCgroups *bool` field to the `SecurityContext` struct in the core v1 API. When set to `true` on cgroup v2 systems, this field will instruct the container runtime to mount the cgroup filesystem with write permissions for the container's own cgroup subtree.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: writable-cgroups-example
+spec:
+  containers:
+  - name: app
+    image: myapp:latest
+    securityContext:
+      writableCgroups: true
```

Comment:
instead of this API, how would you feel about creating a new volume type? it could be type `Cgroup` and only accept the `readOnly` field. If not specified, the container would get a cgroup normal, but if `Cgroup.readOnly` is set to true, then that'll be passed down to the CRI. that also gives us room to customize the cgroup mount more in the future (not sure how we would, but more extendable than the boolean we have today). thoughts?

Reply:
It's an interesting idea, but I have concerns that modeling this as a Volume might confuse users, for example, how do we handle the mount points? Do we restrict it to only `/sys/fs/cgroup`?

I think semantically this feels closer to `readOnlyRootFilesystem` or `procMount`, and represents a modification to the container's isolation boundary, rather than provisioning storage.

To keep things extensible, we could make the security context field a struct or enum instead of a boolean like `procMount`? What do you think?

Reply:
yeah let's go enum. Maybe ``cgroupMountMode: writable/readOnly` to start